src/transformers/generation/continuous_batching.py

```diff
@@ -0,0 +1,408 @@
+        self.cumulative_seqlens_k  = self.cumulative_seqlens_k[keep_mask]
+
+        if evict_mask.sum(-1) > 0:
+            evict_mask = torch.where(evict_mask)[0]        # delete the cache positions for these tokens
+            evict_mask = evict_mask.tolist()
+            self.cache_index = [self.cache_index[k] for k in range(len(self.cache_index)) if k not in evict_mask]
+            for k in evict_mask:
+                cache.free_blocks += cache.block_table[k]
+                del cache.block_table[k]
+                self.finished_sequences.append( self.prompts[k] + self.generated_ids[k])
+                del self.generated_ids[k]
+                del self.prompts[k]
+
+
+    def __len__(self):
+        return len(self.input_tokens)
+
+class ContinuousMixin:
+
+    @torch.no_grad()
+    def fast_generate(
+        self,
+        input_tokens: list = None,
+        logits_processor: Optional[LogitsProcessorList] = None, # TODO
+        stopping_criteria: Optional[StoppingCriteriaList] = None, # TODO
+        synced_gpus: Optional[bool] = None, # TODO
+        streamer: Optional["BaseStreamer"] = None, # TODO
+        **kwargs,
+    ):
+        r"""
+            Fast generate that leverages continuous batching.
+
+            1. You prepare the inputs
+            2. You run forward pass
+            3. You check for eos:
+                a. Evict requests that are done
+                b. Add new requests
+            4. Pack finished sequences to return them
+            5. Go to step 2.
+
+            NO ATTENTION MASKS
+            NO ASSISTANT MODEL
+
+        """
+        self.generation_config.batch_size = 8
+        self.generation_config.max_num_blocks_per_seq = 16
+        self.generation_config.eos_token_id = 1
+
+        input_tokens = [k["input_ids"] for k in input_tokens]
+        paged_attention_cache = PagedAttentionCache(self.config, input_tokens, self.generation_config, self.device) # init with throughput desiered, batch, etc
+        current_batch = ContinuousBatch(input_tokens, self.generation_config)
+
+        while len(current_batch) > 0:
+            continous_batch, position_ids, cache_index, fill_index = current_batch.prepare_next_inputs(paged_attention_cache)
+            kwargs = {
+                "cumulative_seqlens_q":current_batch.cumulative_seqlens_q,
+                "cumulative_seqlens_k":current_batch.cumulative_seqlens_k,
+                "max_seqlens_q":current_batch.max_seqlens_q,
+                "max_seqlens_k":current_batch.max_seqlens_k,
+                "cache_index": cache_index,
+                "fill_index": fill_index,
+                "cache":paged_attention_cache,
+            }
+            out = self.model.forward(
+                continous_batch, position_ids=position_ids,  **kwargs
+            ).last_hidden_state
+            logits = self.lm_head(out[:,current_batch.cumulative_seqlens_q, :])
+            # we don't sample for now :)
+            logits = torch.softmax(logits, dim=-1)
```

Comment:
do you need the softmax here? subsequent argmax will return the same index with or without it

Reply:
For sampling yeah normally you do softmax